r/R/type.R

```diff
@@ -40,7 +40,8 @@ DataType <- R6Class("DataType",
     fields = function() {
       DataType__fields(self)
     },
-    export_to_c = function(ptr) ExportType(self, ptr)
+    export_to_c = function(ptr) ExportType(self, ptr),
+    code = function() call2("unknown_type", self$ToString())
```

Comment:
Just out of curiosity, what us this bit doing here with "unknown_type"?

Reply:
Probably no longer needed. 

It's just so that when it's not implemented for a type, at least you get something. Perhaps that can be replaced by a call to `abort()` instead of giving a call to a function that does not exist ?

Reply:
There are a bunch of types we don't have R functions for, e.g. `DECIMAL256`, `MAP` ... but then I'm not sure how I would test it, because I would need to get one of them in the first place ...

Reply:
So I added `arrow:::DayTimeInterval__initialize()` to make a `DayTimeInterval` type, which don't have an R function yet. Bit artificial, but then we get to be able to test what it looks like when we error, e.g. 

``` r
library(arrow, warn.conflicts = FALSE)
#> See arrow_info() for available features

try_ <- function(code) tryCatch(code, error = function(cnd) print(cnd, backtrace = FALSE))

try_(
  arrow:::DayTimeInterval__initialize()$code()
)
#> <error/rlang_error>
#> Error in `code()`: Unsupported type: <day_time_interval>.
#> ℹ The arrow package currently has no function to make these.

try_(
  schema(x = int32(), y = arrow:::DayTimeInterval__initialize())$code()
)
#> <error/rlang_error>
#> Error in `code()`:
#>   Problem getting code for field "y".
#> Caused by error in `code()`:
#>   Unsupported type: <day_time_interval>.
#>   ℹ The arrow package currently has no function to make these.
try_(
  struct(a = arrow:::DayTimeInterval__initialize())$code()
)
#> <error/rlang_error>
#> Error in `code()`:
#>   Problem getting code for field "a".
#> Caused by error in `code()`:
#>   Unsupported type: <day_time_interval>.
#>   ℹ The arrow package currently has no function to make these.
try_(
  list_of(arrow:::DayTimeInterval__initialize())$code()
)
#> <error/rlang_error>
#> Error in `code()`:
#>   Problem getting code for list type.
#> Caused by error in `code()`:
#>   Unsupported type: <day_time_interval>.
#>   ℹ The arrow package currently has no function to make these.
try_(
  large_list_of(arrow:::DayTimeInterval__initialize())$code()
)
#> <error/rlang_error>
#> Error in `code()`:
#>   Problem getting code for large list type.
#> Caused by error in `code()`:
#>   Unsupported type: <day_time_interval>.
#>   ℹ The arrow package currently has no function to make these.
try_(
  fixed_size_list_of(arrow:::DayTimeInterval__initialize(), list_size = 4L)$code()
)
#> <error/rlang_error>
#> Error in `code()`:
#>   Problem getting code for fixed size list type.
#> Caused by error in `code()`:
#>   Unsupported type: <day_time_interval>.
#>   ℹ The arrow package currently has no function to make these.
try_(
  schema(x = int32(), y = struct(a = arrow:::DayTimeInterval__initialize()))$code()
)
#> <error/rlang_error>
#> Error in `code()`:
#>   Problem getting code for field "y".
#> Caused by error in `code()`:
#>   Problem getting code for field "a".
#> Caused by error in `code()`:
#>   Unsupported type: <day_time_interval>.
#>   ℹ The arrow package currently has no function to make these.
```

<sup>Created on 2021-11-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

Reply:
The bit we had previously with an artificial call to `unknown_type()` allowed to have almost all the schema, with just these as placeholders: 

``` r
library(arrow, warn.conflicts = FALSE)
#> See arrow_info() for available features

try_ <- function(code) tryCatch(code, error = function(cnd) print(cnd, backtrace = FALSE))

try_(
  schema(x = int32(), y = struct(a = arrow:::DayTimeInterval__initialize()))$code()
)
#> schema(x = int32(), y = struct(a = unknown_type("day_time_interval")))
```

<sup>Created on 2021-11-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

Reply:
Another way that we could test types that we don't have functions for (though that's really only useful insofar as we are creating skipped tests with jiras to implement them...) would be to create an object with a schema and then cast values to the types that we don't have direct support for.